### PR TITLE
create cloud credential from password fields of node template 

### PR DESCRIPTION
**Pull request by kinarashah**
Cloud credential would have different `credentialConfig` corresponding to configs in node template and would contain password field of the config. 

https://github.com/rancher/rancher/issues/15696

### CODE REVIEW

File: pkg/api/server/managementstored/setup.go [hunk 1]
```diff
@@ -285,9 +286,10 @@ func NodeTemplates(schemas *types.Schemas, management *config.ScaledContext) {
 	s := &nodeTemplateStore.Store{
 		Store:          userscope.NewStore(management.Core.Namespaces(""), schema.Store),
 		NodePoolLister: npl,
+		CredLister:     management.Core.Secrets("").Controller().Lister(),
 	}
 	schema.Store = s
-	schema.Validator = nodetemplate.Validator
+	//schema.Validator = nodetemplate.Validator
```
Comment (alena1108) [inline pkg/api/server/managementstored/setup.go]: shall we remove the commented line?

File: pkg/api/store/nodetemplate/store.go [hunk 2]
```diff
@@ -24,3 +31,39 @@ func (s *Store) Delete(apiContext *types.APIContext, schema *types.Schema, id st
 	}
 	return s.Store.Delete(apiContext, schema, id)
 }
+
+func (s *Store) Create(apiContext *types.APIContext, schema *types.Schema, data map[string]interface{}) (map[string]interface{}, error) {
+	driver := configfield.GetDriver(data)
+	if driver == "" {
+		id := convert.ToString(data["credentialId"])
+		if id == "" {
+			return nil, httperror.NewAPIError(httperror.MissingRequired, "a Config field or credentialId must be set")
```
Comment (alena1108) [inline pkg/api/store/nodetemplate/store.go]: is `Config` the exact field name here? if not, we should somehow better reflect it

File: pkg/controllers/management/drivers/nodedriver/machine_driver.go [hunk 3]
```diff
@@ -136,6 +166,35 @@ func (m *Lifecycle) download(obj *v3.NodeDriver) (*v3.NodeDriver, error) {
 	if err != nil && !errors.IsAlreadyExists(err) {
 		return obj, err
 	}
+	obj, err = m.createCredSchema(obj, credFields)
+	if err != nil {
+		return obj, err
+	}
+
+	//passwordStore.SetPasswordStore(m.schemas, m.secretStore, m.nsStore)
```
Comment (alena1108) [inline pkg/controllers/management/drivers/nodedriver/machine_driver.go]: should we remove it?

File: pkg/controllers/management/drivers/nodedriver/machine_driver.go [hunk 4]
```diff
@@ -136,6 +166,35 @@ func (m *Lifecycle) download(obj *v3.NodeDriver) (*v3.NodeDriver, error) {
...
+	credentialSchema.Name = obj.Spec.DisplayName + "credentialconfig"
+	credentialSchema.OwnerReferences = []metav1.OwnerReference{
+		{
+			UID:        obj.UID,
+			Kind:       obj.Kind,
+			APIVersion: obj.APIVersion,
+			Name:       obj.Name,
+		},
+	}
+	_, err := m.schemaClient.Create(credentialSchema)
```
Comment (alena1108) [inline pkg/controllers/management/drivers/nodedriver/machine_driver.go]: is that easier to always create it, and check for the error, vs fetching the object first to check if it exists?

File: pkg/controllers/management/node/controller.go [hunk 5]
```diff
@@ -132,20 +140,33 @@ func (m *Lifecycle) Create(obj *v3.Node) (runtime.Object, error) {
...
 		if !ok {
 			return obj, fmt.Errorf("node config not specified")
 		}
 		if template.Spec.Driver == amazonec2 {
 			setEc2ClusterIDTag(rawConfig, obj.Namespace)
 		}
-
+		credID := convert.ToString(values.GetValueN(data, "spec", "credentialName"))
+		if credID != "" {
+			existingSchema, err := m.schemaLister.Get("", template.Spec.Driver+"config")
```
Comment (alena1108) [inline pkg/controllers/management/node/controller.go]: can we perhaps extract the code in if block to its own separate method?

File: app/machinedriver_data.go [hunk 6]
```diff
@@ -42,22 +48,31 @@ func addMachineDrivers(management *config.ManagementContext) error {
 		"c31b9da2c977e70c2eeee5279123a95d", []string{"ecs.aliyuncs.com"}, false, false, management); err != nil {
 		return err
 	}
-
-	return addMachineDriver("vmwarevsphere", "local://", "", nil, true, true, management)
+	return nil
```
Comment (alena1108) [inline app/machinedriver_data.go]: why did we remove the vmwaresphere driver?

File: app/machinedriver_data.go [hunk 7]
```diff
@@ -42,22 +48,31 @@ func addMachineDrivers(management *config.ManagementContext) error {
...
 }
 
 func addMachineDriver(name, url, checksum string, whitelist []string, active, builtin bool, management *config.ManagementContext) error {
 	lister := management.Management.NodeDrivers("").Controller().Lister()
 	cli := management.Management.NodeDrivers("")
 	m, _ := lister.Get("", name)
+	annotations := map[string]string{}
```
Comment (alena1108) [inline app/machinedriver_data.go]: can we elaborate in the pr description on how annotations are used (to identify the credential filed, etc etc)

File: pkg/controllers/management/drivers/nodedriver/machine_driver.go [hunk 8]
```diff
@@ -57,12 +66,31 @@ func (m *Lifecycle) download(obj *v3.NodeDriver) (*v3.NodeDriver, error) {
...
+		for name, field := range existingSchema.Spec.ResourceFields {
+			if pwdFields[name] {
+				field.Type = "password"
+			}
+			if field.Type == "password" || userCredFields[name] {
+				credFields[name] = field
+			}
+		}
+		_, err := m.createCredSchema(obj, credFields)
+		if err != nil {
```
Comment (alena1108) [inline pkg/controllers/management/drivers/nodedriver/machine_driver.go]: nit: just `return obj, err`

File: pkg/controllers/management/drivers/nodedriver/machine_driver.go [hunk 9]
```diff
@@ -136,6 +173,37 @@ func (m *Lifecycle) download(obj *v3.NodeDriver) (*v3.NodeDriver, error) {
 	if err != nil && !errors.IsAlreadyExists(err) {
 		return obj, err
 	}
+	obj, err = m.createCredSchema(obj, credFields)
```
Comment (alena1108) [inline pkg/controllers/management/drivers/nodedriver/machine_driver.go]: nit: return m.createCredSchema

File: pkg/controllers/management/drivers/nodedriver/machine_driver.go [hunk 10]
```diff
@@ -136,6 +173,37 @@ func (m *Lifecycle) download(obj *v3.NodeDriver) (*v3.NodeDriver, error) {
...
+	if err != nil {
+		return obj, err
+	}
+
+	return obj, nil
+}
+
+func (m *Lifecycle) createCredSchema(obj *v3.NodeDriver, credFields map[string]v3.Field) (*v3.NodeDriver, error) {
+	name := obj.Spec.DisplayName + "credentialconfig"
```
Comment (alena1108) [inline pkg/controllers/management/drivers/nodedriver/machine_driver.go]: just to confirm: is it ok the schema name to be lowercased? `credentialconfig`. Also check if display name of the node driver can be replaced with the actual schema name of the node driver

File: pkg/controllers/management/drivers/nodedriver/machine_driver.go [hunk 11]
```diff
@@ -151,6 +219,10 @@ func (m *Lifecycle) Updated(obj *v3.NodeDriver) (runtime.Object, error) {
 		return obj, err
 	}
 
+	if err := m.createOrUpdateNodeForEmbeddedTypeCredential(obj.Spec.DisplayName+"credentialconfig", obj.Spec.DisplayName+"credentialConfig", obj.Spec.Active); err != nil {
```
Comment (alena1108) [inline pkg/controllers/management/drivers/nodedriver/machine_driver.go]: we should also extract name generation to its own method as it's used multiple times.

File: pkg/controllers/management/node/controller.go [hunk 12]
```diff
@@ -29,6 +32,7 @@ import (
 const (
 	defaultEngineInstallURL = "https://releases.rancher.com/install-docker/17.03.2.sh"
 	amazonec2               = "amazonec2"
+	credentialNS            = "cattle-global-data"
```
Comment (alena1108) [inline pkg/controllers/management/node/controller.go]: I feel like we already use this namespace name in several places; can you define the global ns in some common util file, and ref from there in all the places

File: pkg/httpproxy/sign.go [hunk 13]
```diff
@@ -0,0 +1,235 @@
...
+	"strings"
+	"time"
+
+	"github.com/aws/aws-sdk-go/aws/credentials"
+	"github.com/aws/aws-sdk-go/aws/signer/v4"
+	"github.com/rancher/types/apis/core/v1"
+)
+
+const (
+	CredentialNamespace = "cattle-global-data"
```
Comment (alena1108) [inline pkg/httpproxy/sign.go]: same here - define global namespace in a common util

File: pkg/httpproxy/proxy.go [hunk 14]
```diff
@@ -118,5 +119,15 @@ func (p *proxy) proxy(req *http.Request) error {
...
 	req.Header = headerCopy
 
+	if auth != "" {
+		req.Header.Set("Authorization", auth)
+	} else if cAuth != "" {
+		signer := newSigner(cAuth)
+		if signer != nil {
+			return signer.sign(req, p.credentials, cAuth)
+		}
+		req.Header.Set("Authorization", cAuth)
```
Comment (alena1108) [inline pkg/httpproxy/proxy.go]: can we reference auth header? as it's defined in https://github.com/rancher/rancher/pull/17014/files#diff-dbac85ab3314162046444898deb8e4bbR25

File: pkg/httpproxy/proxy.go [hunk 15]
```diff
@@ -118,5 +119,15 @@ func (p *proxy) proxy(req *http.Request) error {
 	req.URL = destURL
 	req.Header = headerCopy
 
+	if auth != "" {
```
Comment (alena1108) [inline pkg/httpproxy/proxy.go]: can we put some comments on what the split between auth and cAuth means

**Comment by alena1108** [comment]
Do not merge till we figure out rbac for cloud credentials

File: pkg/controllers/management/drivers/nodedriver/machine_driver.go (hunk 10 above)
Comment (kinarashah) [reply pkg/controllers/management/drivers/nodedriver/machine_driver.go]: yeah, it's ok, display name of node driver is created from the node driver in a similar way, 
`dynamicSchema.Name = obj.Spec.DisplayName + "config"`

File: pkg/controllers/management/cloudcredential/controller.go [hunk 16]
```diff
@@ -0,0 +1,62 @@
...
+	m := Controller{
+		cloudCredentials:  cloudCredentials,
+		managementContext: management,
+	}
+	m.cloudCredentials.AddHandler(ctx, "management-cloudcredential-controller", m.ccSync)
+}
+
+func (n *Controller) ccSync(key string, cloudCredential *v1.Secret) (runtime.Object, error) {
+	if cloudCredential == nil || cloudCredential.DeletionTimestamp != nil {
+		return cloudCredential, nil
```
Comment (alena1108) [inline pkg/controllers/management/cloudcredential/controller.go]: could you point to the place where role binding is being removed?

File: pkg/controllers/management/drivers/nodedriver/machine_driver.go [hunk 17]
```diff
@@ -136,6 +169,32 @@ func (m *Lifecycle) download(obj *v3.NodeDriver) (*v3.NodeDriver, error) {
...
 	}
+	return m.createCredSchema(obj, credFields)
+}
+
+func (m *Lifecycle) createCredSchema(obj *v3.NodeDriver, credFields map[string]v3.Field) (*v3.NodeDriver, error) {
+	name := credentialConfigSchemaName(obj.Spec.DisplayName)
+	_, err := m.schemaLister.Get("", name)
+	if err != nil && !errors.IsNotFound(err) {
+		return obj, err
+	} else if errors.IsNotFound(err) {
```
Comment (alena1108) [inline pkg/controllers/management/drivers/nodedriver/machine_driver.go]: what if error is nil here? does IsNotFound handle it?

File: pkg/controllers/management/node/controller.go [hunk 18]
```diff
@@ -465,3 +474,41 @@ func roles(node *v3.Node) []string {
...
+func (m *Lifecycle) setCredFields(data interface{}, fields map[string]v3.Field, credID string) error {
+	splitID := strings.Split(credID, ":")
+	if len(splitID) != 2 {
+		return fmt.Errorf("invalid credential id %s", credID)
+	}
+	cred, err := m.credLister.Get(namespace.GlobalNamespace, splitID[1])
+	if err != nil {
+		return err
+	}
+	if ans, ok := data.(map[string]interface{}); ok {
```
Comment (alena1108) [inline pkg/controllers/management/node/controller.go]: do you think we could use `convert` utility? for interface to map conversion

File: pkg/controllers/management/drivers/nodedriver/machine_driver.go (hunk 17 above)
Comment (kinarashah) [reply pkg/controllers/management/drivers/nodedriver/machine_driver.go]: thank you! updated

File: pkg/controllers/management/cloudcredential/controller.go (hunk 16 above)
Comment (kinarashah) [reply pkg/controllers/management/cloudcredential/controller.go]: https://github.com/rancher/rancher/blob/7fe9051d857866bd70291a8f026634b76b9e5cd1/pkg/controllers/management/globalnamespacerbac/rbac_common.go#L142

**Comment by alena1108** [comment]
LGTM

File: pkg/httpproxy/sign.go [hunk 19]
```diff
@@ -0,0 +1,230 @@
...
+	}
+	return nil
+}
+
+func (br bearer) sign(req *http.Request, secrets v1.SecretInterface, auth string) error {
+	data, secret, err := getAuthData(auth, secrets, []string{"passwordField", "credID"})
+	if err != nil {
+		return err
+	}
+	req.Header.Set(AuthHeader, fmt.Sprintf("%s %s", "Bearer", secret[data["passwordField"]]))
```
Comment (prachidamle) [inline pkg/httpproxy/sign.go]: Is the secret[data["passwordField"]] encoded?

File: pkg/httpproxy/sign.go [hunk 20]
```diff
@@ -0,0 +1,230 @@
...
+	}
+	resp, err := doNewRequest(req) // request to get challenge fields from server
+	if err != nil {
+		return err
+	}
+	challengeData, err := parseChallenge(resp.Header.Get("WWW-Authenticate"))
+	if err != nil {
+		return err
+	}
+	logrus.Debugf("digest challenge data %v", challengeData)
```
Comment (prachidamle) [inline pkg/httpproxy/sign.go]: Hope we dont log any sensitive info here

Comment (kinarashah) [reply pkg/httpproxy/sign.go]: it's the details (nonce,  etc) server responds which we then use for authenticate request, this usually should be new each time but I'll remove it to be on safer side

File: pkg/httpproxy/sign.go [hunk 21]
```diff
@@ -0,0 +1,230 @@
...
+	"github.com/rancher/types/apis/core/v1"
+)
+
+type Signer interface {
+	sign(*http.Request, v1.SecretInterface, string) error
+}
+
+func newSigner(auth string) Signer {
+	splitAuth := strings.Split(auth, " ")
+	switch strings.ToLower(strings.ToLower(splitAuth[0])) {
```
[the remark this answers is not in the document]
Comment (kinarashah) [reply pkg/httpproxy/sign.go]: thank you!! removed

File: pkg/httpproxy/sign.go (hunk 19 above)
Comment (kinarashah) [reply pkg/httpproxy/sign.go]: no

**Comment by prachidamle** [comment]
LGTM

**Pull request merged**